### ISSUE TITLE
Speed up front container startup by pre-compiling

### ISSUE THREAD

**Issue by moufmouf** [thecodingmachine/workadventure#594]
Compiling is currently done at container startup because we need to take into account the API_URL environment variable (that is compiled into JS).
But this takes about 40 seconds, which is 40 seconds of downtime.

We could simply search and replace the string in the JS but that would break the sourcemap.

Hopefully, there is this little utility: https://www.npmjs.com/package/sourcemap-aware-replace

We could build the JS directly in the container and only search and replace the API_URL string in the files at container startup.

\o/

**Comment by Informatic** [comment]
Wouldn't #577 be somewhat relevant here? You could also inject configuration into `index.html` on startup. (eg. by exposing it in global `window` object before loading main javascript bundle)

**Comment by priegger** [comment]
From my understanding the current code expects the environment variables to be declared during build, then packages the values with the application an injects them during startup.

How about allowing the user to override variables for pre-built images? A (mostly) backwards compatible system could work like this:

- Enironment variables specified during build are still packaged but renamed (e.g. a _DEFAULT suffix is added)
- Environment variables specified during container startup are used, if they are not specified, the _DEFAULT versions are used

What do you think about this? I think the breaking change here is, that if the code is used and built directly (so the docker images are not used) and environment variables like the API_URL are specified during start, those values are ignored and the values specified during build are used.

Is there a list of environment variables/settings that can be specified during container start and that need to be supported in this solutions? Would you accept a patch for this, the environment variable handling refactoring and the docker container image with the pre-built application?

**Comment by jonnytest1** [comment]
it should speed up the entire thing quite a bit if you extract the environment vars into its own module

**Comment by LittleFox94** [comment]
I'm annoyed enough by the current behavior to want to do some useful changes ^^

At first I thought, only the GA tracking is the culprit and wanted to have a JS file which is always loaded by the HTML, empty by default, but optionally gets filled on startup - with all the application already built.

I think this should also work with everything else of the frontend - having that completely built and a `config.generated.js` loaded by the HTML, before the App is loaded. That `config.generated.js` would contain the GA token and `API_URL` (or `PUSHER_URL`). Anything else I'd have to take care of?

For GA there is then the problem, currently having a custom HTML snippet is allowed, too - this would not be possible that way .. is this known to be used?

I would also like to change the base image to nginx or any other static file webserver, not dragging in PHP like it's done now - oh and make it a multi-stage dockerfile

Comments on that would be great, especially on GA custom HTML snippet requirement and if that approach seems to make sense.

**Comment by victorgetz** [comment]
This is a big issue in my point of view.

Autoscaling with kubernetes does not working properly because
- The startup time takes ages for the frontend -> we need to set the cpu/memory upscale threshold very low 
- The ressource consumption is super huge at startup and after the build it takes just a percentage of CPU/Memory -> ressource quotes definitions are hard to define

**Comment by victorgetz** [comment]
I currently use a hack in Kubernetes where i use a pod which will build me upfront the dist folder and then inject it into the other containers via persistent volume.

`
apiVersion: v1
kind: Pod
metadata:
  name: {{ .Chart.Name }}-build-frontend
  labels:
    app: {{ .Chart.Name }}-build-frontend
spec:
  containers:
    - name: {{ .Chart.Name}}-build-frontend
      image: "{{ .Values.front.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
      imagePullPolicy: {{ .Values.front.image.pullPolicy }}
      command: [ "/bin/sh","-c" ]
      args: ["cp dist/index.tmpl.html dist/index.tmpl.html.tmp && yarn run build && rm -Rf /tmp/dist/* && mv dist/* /tmp/dist"]
      env:
        - name: 'API_URL'
          value: {{ include "workadventure.pusherUrl" . }}
        - name: 'START_ROOM_URL'
          value: "{{- .Values.startRoomURL -}}"
        - name: 'JITSI_URL'
          value: "{{- .Values.jitsi.url -}}"
        - name: 'SECRET_JITSI_KEY'
          value: "{{- .Values.jitsi.secretKey -}}"
        - name: 'JITSI_PRIVATE_MODE'
          value: "{{- .Values.jitsi.privateMode -}}"
        - name: "WEBEX_GLOBAL_SPACE_ID"
          value: "{{- .Values.webex.GlobalSpaceId -}}"
        {{- if .Values.front.extraEnvVars }}
        {{- include "common.tplvalues.render" (dict "value" .Values.front.extraEnvVars "context" $) | nindent 8 }}
      {{- end }}
      volumeMounts:
        - mountPath: /tmp/dist
          name: {{.Values.pvc.volumeName}}
  restartPolicy: Never
  volumes:
    - name: {{.Values.pvc.volumeName}}
      persistentVolumeClaim:
        claimName: {{.Values.pvc.volumeName}}

`

**Comment by anbraten** [comment]
Injecting environment variables into compiled JS code is a common problem I saw in many projects. One of the easiest solution I saw is IMO to load an additional file like the suggested `config.generated.js` which contains all needed environment variables. In dev mode this route could be served by webpack and in production it can simply be generated by a shell script replacing some placeholders with the corresponding env vars.
